authentication-via-social-networks/README.md

```diff
@@ -0,0 +1,726 @@
+
+  belongs_to :user
+
+  validates :user, :provider, :uid, presence: true
+  validates :uid, uniqueness: { scope: :provider }
+
+  def self.from_omniauth(auth)
+    find_by(provider: auth.provider, uid: auth.uid)
+  end
+end
+```
+
+```ruby
+# app/models/user.rb
+class User < ActiveRecord::Base
+  devise ..., :omniauthable, omniauth_providers: Identity::PROVIDERS
+
+  has_many :identities, dependent: :destroy
+  ...
+end
+```
+
+```ruby
+# app/policies/auth_verification_policy.rb
+class AuthVerificationPolicy
+  class OauthError < StandardError
+  end
+
+  attr_reader :auth
+  private :auth
+
+  def initialize(auth)
+    @auth = auth
+  end
+
+  def verified?
+    send(auth.provider)
+  rescue NoMethodError
+    fail OauthError, I18n.t("omniauth.verification.not_implemented", kind: auth.provider)
+  end
+
+  private
+
+  def facebook
+    auth.info.verified? || auth.extra.raw_info.verified?
+  end
+
+  def google_oauth2
+    auth.extra.raw_info.email_verified?
+  end
+end
+```
+
+```ruby
+# app/views/identities/_list.html.slim
+- if identities.any?
+  b Successfully authorized via:
+  ul.js-identities
+    - identities.each do |identity|
+      li = render "identities/provider_link", identity: identity
+b Add service to sign in with:
+ul
+  - Identity::PROVIDERS.each do |provider|
+    li = link_to provider_name(provider), user_omniauth_authorize_path(provider)
+```
+
+```ruby
+# app/views/identities/_provider_link.html.slim
+= link_to "#{provider_name(identity.provider)} (#{identity.uid.truncate(9)}). Unauthorize?",
+          identity,
+          data: { confirm: "Are you sure you want to remove this identity?" },
+          method: :delete,
+          class: "js-unauthorize"
+```
+
+```ruby
+# app/views/users/registrations/edit.html.slim
+...
+= render "identities/list", identities: current_user.identities
+...
+```
+
+```ruby
+# config/initializers/devise.rb
+...
+config.omniauth :google_oauth2, ENV["GOOGLE_CLIENT_ID"], ENV["GOOGLE_CLIENT_SECRET"]
+config.omniauth :facebook, ENV["FACEBOOK_APP_ID"], ENV["FACEBOOK_APP_SECRET"], info_fields: "email, name, verified"
+...
+```
+
+```ruby
+# config/locales/models/identity.en.yml
+en:
+  active_record:
+    attributes:
+      identity:
+        provider_name:
+          google_oauth2: Google
+          facebook: Facebook
+```
+
+```ruby
+# config/locales/oauth.en.yml
+en:
+  omniauth:
+    verification:
+      failure: Please confirm your %{kind} account before continuing.
+      not_implemented: Verification checking is not implemented for %{kind}.
+```
+
+```ruby
+# config/routes.rb
+...
+devise_for :users, controllers: { omniauth_callbacks: "omniauth_callbacks" }
+resources :identities, only: :destroy
+...
+```
+
+```ruby
+# db/migrate/20160531141911_create_identities.rb
+class CreateIdentities < ActiveRecord::Migration
+  def change
+    create_table :identities do |t|
+      t.belongs_to :user, index: true
+      t.string :provider, index: true, null: false, default: ""
```

Comment:
Value of `provider` and `uid` columns should never be empty string. Let's get rid of default values for these columns.